### ISSUE TITLE
[Questions/Bugs] Peers aren't shown & Upload stuck on 0kb/s & Download not stable

### ISSUE THREAD

**Issue by burekas7** [elgatito/plugin.video.elementum#55]
1) Why there are no Peers anymore in the Overlay Status? (There are only Seeds)

2) Why the upload rate is stuck on 0kb/s while downloading when I set maximum upload rate in the settings to 500kb/s for example (Is it connect to the issue that there are no peers anymore?)
Is it should be like that?

3) What could be the reason that the download speed is not so stable somtimes (Jumps 1M, 4M, 2M, 700k, 3M etc...)
- Also I caught it one time that it show a negative Download speed (-300kb/s) .

## My Environment
* Version used: Elementum v0.0.30
* Environment name and version: Kodi 17.6 [Krypton]
* Operating System and version: Windows 10, 64bit

**Comment by elgatito** [comment]
You should give more information, which storage type you use.
In general:
1) We use information from the torrent library. It does not split into peers and seeds. 
2) Upload is currently disabled for memory storage
3) For memory you don't download all file but only pieces you need. That is why speed can go up and down until it downloads what it needs.
4) When piece of file is not needed anymore - its' deleted and size of downloaded data is decreasing, which means negative numbers.

**Comment by burekas7** [comment]
@elgatito 
* File system: NTFS (The same as I used for Quasar)

1,2) But let me understand something, this is a fork of Quasar.
In Quasar there is info separated for Seeders and Peers, so why is it different here?
And also regarding the uploading, what is the difference here from Quasar? (This is the same memory storage)
In general, what have you changed here from Quassar regarding the download/upload system?

**Comment by elgatito** [comment]
Elementum and Quasar have different torrent downloading engines. All download/upload is performed by engines, not addons. 
So any issuds regrdinv bad speed or something like that, are not applicable to this addon.

**Comment by burekas7** [comment]
So regarding this type of file system and OS, is it normal that the upload is staying on 0kb/s even though I put their a speed limit? Can this affect on my download process

By the way can you see something in the log or is it irrelevant?

**Comment by elgatito** [comment]
Upload limits do not affect downloading. Anyway, if you have issues with download/upload speed, you can create a concrete issue with a log and http://KODI_IP:65220/info content, showing health of a torrent engine. Then, if it looks like a torrent engine problem - issue will be forwarded to the author of a library.

**Comment by burekas7** [comment]
OK. I will check it again and will take the relevant info.
But I still don't understand why have you needed to change the torrent engine, because it is working well in Quasar. There were only some GUI problems that remained open.

**Comment by burekas7** [comment]
Basically I'm not satisfied with this engine.
Download speed very not stable, there is no info about seeds and peers, timeouts.

Here is the info file:
https://pastebin.com/DWXmRaPE

**Comment by elgatito** [comment]
> But I still don't understand why have you needed to change the torrent engine, because it is working well in Quasar.

1) Because Quasar's engine does not allow custom storages, like memory.
2) Because I wanted.

You are free to fork Quasar, do the changes and use, or use any other fork that keep Quasar engine.

> Download speed very not stable, there is no info about seeds and peers, timeouts.

You should provide kodi.log.

**Comment by burekas7** [comment]
Here is the kodi log: https://pastebin.com/HneqWubJ
Here is the torrent info log: https://pastebin.com/Cc8R7Cte
(Tell me if you need something more)

I have understood that Elementum is using a different torrent engine, but I still don't figure out why the upload rate is on 0 kb/s , and why there is no separation between seeds and peers.

The idea with torrents is about sharing and if you have never enable seeding and there is no peers, it's a little problematic. 
And regarding what you said that "Upload is currently disabled for memory storage", is it going to be fixed soon? Also the developer of the torrent engine can maybe consider of add the option of showing the seeds and peers separately?
* I understood that the engine was changed due to new memory support, but I think it's important to check this.

And thank you for your great work here, I'm very appreciate that.
I'm just trying to help and improve, I'm not against you or something.
@elgatito

**Comment by burekas7** [comment]
* Quasar:
![quasar2](https://user-images.githubusercontent.com/13424257/35189411-ea40e688-fe41-11e7-9db3-a02b78323e8a.png)

* Elementum - Upload is on 0 kb/s and where are all the peers? It can improve the download speed and also the 'sharing' issue.
![elementum1](https://user-images.githubusercontent.com/13424257/35189410-e9f52b76-fe41-11e7-944a-755299635514.png)

**Comment by elgatito** [comment]
> And regarding what you said that "Upload is currently disabled for memory storage", is it going to be fixed soon? 

Maybe in future.

>Also the developer of the torrent engine can maybe consider of adding the option of showing the seeds and peers separately?

You are free to raise a request in torrent engine repository for that - https://github.com/anacrolix/torrent
Again, we are using what we have from torrent engine.

**Comment by herrwusel** [comment]
>> And regarding what you said that "Upload is currently disabled for memory storage", is it going to be fixed soon?
> Maybe in future.

Is that depending on the torrent engine or on Elementum? I agree that sharing is caring. Especially for private trackers.

**Comment by elgatito** [comment]
When memory storage is stable for reading, we can try enabling seeding.

**Comment by burekas7** [comment]
I think something goes bad in 0.0.34
Download rate is very low (below 500kbs), also the download suddenly fell to zero and stuck there, while the seeds count was continue to be updated and after a while they fell to few seeds only.
* Kodi log: 
Part 1 - https://pastebin.com/5GZY3UaG
Part 2 - https://pastebin.com/yFtqWBh3
* Torrent engine log: https://pastebin.com/4APwd29C

By the way the upload rate is still stuck in 0kbs and there is no peers yet.

**Comment by elgatito** [comment]
> Download rate is very low (below 500kbs)

Try without download speed limit and see if it helps.

**Comment by burekas7** [comment]
Still not helps, even worse I think, but pretty the same problem.
Download rate is below 700kbs , it is suddenly jump to 1M,2M but then it fell to 0kbs and stuck there and showing "Queued" status.

* Before (With setting the download limit of 8000kbs):
Kodi Log:
Part 1 - https://pastebin.com/G4GjaKct
Part 2 - https://pastebin.com/Jnc3rhFY
Torrent Log: 
https://pastebin.com/HTvLrkbR

After (Without setting the download limit)
Kodi Log:
https://pastebin.com/QbxWZyF8
Torrent Log: 
https://pastebin.com/ZKF74fbJ

**Comment by burekas7** [comment]
Is there any update regarding this issue (I have tested it also with 0.0.35)

By the way, why the active seeds amount is always stuck on max 80?
Log: https://pastebin.com/PUP0vw4p

**Comment by elgatito** [comment]
> By the way, why the active seeds amount is always stuck on 80 maximum?

Default limit is set to 80. Not every system can handle more connections.

**Comment by burekas7** [comment]
OK there is a little improvement with version 0.0.36 regarding the download speed (still can be improved)
But there are some problems with the initial buffer.
* It usually start direct from ~11% and not from 0%
* When it reached to ~99% of the buffer, although it had continued downloading, it stucked on this percentage too long (~30 seconds) and then start to play the video.

Kodi Log:
[v36KodiLog.txt](https://github.com/elgatito/plugin.video.elementum/files/1671275/v36KodiLog.txt)

***
By the way, can I change the limit of 80 in the "Connection Limit"?
And when do you think the Upload will begin to work and also the Peers using in addition to the seeds?

**Comment by elgatito** [comment]
We show what we get from the torrent library. 
If it's stuck on 99% then it waits for the last chunks to be downloaded.

> By the way, can I change the limit of 80 in the "Connection Limit"?

Change connection limit in Bittorrent tab of settings.

**Comment by burekas7** [comment]
How can I know that the number of connection begin to make problems?
The internet connection become slow or something? What do you mean by 'not every ystem can handle it'?

**Comment by elgatito** [comment]
Example: RPI 1 need lower connection limit as it is killing the system.

**Comment by burekas7** [comment]
Ok, so what is the indication that could tell me the maximum recommended connection limit for my system?

**Comment by elgatito** [comment]
- Change storage engine to memory
- Set "buffer" in "Bittorrent" tab to 100mb
- Start any torrent and check the maximum download speed while buffer is downloaded. 
- If that is low - you have some problem, if not - then it's all about file storage speed.

**Comment by burekas7** [comment]
I have another questions regarding the seeds comparing to Quasar: (Regarding the images below)
1. Why there is a huge difference in the numbers of the the seeds between Quasar and Elementum?
(This is the same torrent)
2. Is the max seeds in quasar is meaning the same as here, or is it a different meaning of 'Number of connections'?

Quasar:
![quasar1](https://user-images.githubusercontent.com/13424257/35485539-1ad59f10-0459-11e8-9a4c-b6694ce1ef49.png)

Elementum:
![elementum1](https://user-images.githubusercontent.com/13424257/35485537-121604fa-0459-11e8-8c42-edb7a62e84c7.png)

**Comment by burekas7** [comment]
Updates (or not so update)
Download is still not so stable but better then previous version, it move between 800kbs to 1800kbs with 120 connections.

But there is a new problem, as I said before it stuck on ~99% for a while, but this time Elementum get timeout and show "Playback fails". After that the download still continued and reached to 100% then nothing happen, and the download still in progress in the background. I tried to click again and chose the same torrent, but something went wrong after that, and the download stopped.

Kodi Log: https://paste.ubuntu.com/26485049/
Torrent log (On Buffer): https://paste.ubuntu.com/26485061/

By the way, is there any progress regarding the upload and the peers?

And also it would be nice if you write on the release page what has been updated on each version and what are the related issues of each fix (If there is). Thanks.

**Comment by elgatito** [comment]
> Download is still not so stable but better then previous version, it move between 800kbs to 1800kbs with 120 connections.

If you set download storage to memory and the download goes much faster then with files - then you have issues with processing files. 
That is OS specific usually, and not controlled by this plugin.


> And also it would be nice if you write on the release page what has been updated on each version and what are the related issues of each fix (If there is). Thanks.

There is a Changelog menu item in the main menu of Elementum, which shows some descent updates in the plugin.

**Comment by burekas7** [comment]
But in quasar I also using file storage and the speed is much better then here
(The memory storage doesn't save the file so I don't like it, but I will use it again to test the speed)
And as I see, that it's not only my problem, other people also report this.

Regarding the Changelog, I know it.
But it will be useful to mention it also in the release page and as I said also to mention which fix relate to which github issue if has.
By the way, on the two last releases you didn't write about them in the Changelog.

**Comment by elgatito** [comment]
File operations are done by torrent engine. This plugin has nothing to do with it.
 
If you want - you can raise an issue there: https://github.com/anacrolix/torrent

**Comment by burekas7** [comment]
I have tried now the memory storage for speed testing, I used 100mb buffer and 120 connections.
The speed reached to ~2200kbs. But after the buffer it get into 'Queued' status, so I don't know how stable it is. But still it slower than Quasar.
Is it normal?

By the way, is there any way to use Elementum addon but use the Quasar torrent engine, or is too complicate?
Because I like the work you are doing with the addon, but less satisfied with the torrent engine at this poine now.
